unix: Refactor variant config and enable features in standard build.

jimmo opened the pull request:
Summary:
 - Remove the special-casing of the minimal variant and simplify mpconfigport/mpconfigvariant.
 - Make the default build useful by default.
 - Remove "unused" variants.
 - Make coverage build use the "everything" level, and update mpconfig.h to define features for that level.

This is a no-op for the coverage variant.

For the minimal variant it is effectively the same except:
 - some additional configuration macros get defined but are not used (e.g. `MICROPY_MACHINE_MEM_GET_READ_ADDR`)
 - three new features macros enabled -- `MICROPY_PY_SYS_EXIT` & `MICROPY_PY_SYS_PATH_ARGV_DEFAULTS` & `MICROPY_MODULE_OVERRIDE_MAIN_IMPORT`. Arguably those last two should have been enabled already.

The dev variant and standard variant are now combined into the standard variant. The logic here is:
- The default unix build should be _at least_ feature compatible with a common bare-metal board. e.g. pybv11. This means enabling asyncio, random extra, etc. The full list is below.
- I know there's history here, but my logic is that the default build is primarily for developers to run MicroPython code locally (either for development/debugging of MicroPython itself, or for rapid iteration of code intended for a device). If people want to tune the Unix port for a specific target, this is now much easier to do with config levels and the simplified mpconfigport/mpconfigvariant layout introduced in this PR.

Removes the "fast" and "freedos" variants. Fast doesn't really make sense except to game a specific benchmark. Perhaps in the future we should replace it with experimental optimisations that we don't have code size for on bare-metal. "freedos" is too difficult to maintain (and isn't tested by CI either).

Bluetooth (which was enabled in the dev variant) is disabled in the default variant. It adds an extra dependency on NimBLE, and isn't terribly useful out-of-the-box because it requires a HCI UART. It can be easily enabled by `make` arguments.

settrace was also enabled in dev. It can also be enabled easily with a single `make` argument, but I would suggest that we enable this by default in the default build.

Another thing I'd like to consider is making the default variant even more like bare metal, e.g. use extmod/uselect rather than unix/select (to ensure compatibility with bare-metal -- e.g. aioble doesn't work with the unix version), and disabling FFI/JNI/etc (to reduce build dependencies). I suggest we could add a "posix" (or "unix"?) variant that enables these "I'm using this as a CPython replacement" use cases.

TODO: Update README (I want to move the Unix section out of the top level and into ports/unix) and docs.

cc @dlech @mattytrentini @andrewleech 

Extra features added to standard:
```
MICROPY_ENABLE_SCHEDULER
MICROPY_MODULE_ATTR_DELEGATION
MICROPY_MODULE_BUILTIN_INIT
MICROPY_OPT_MATH_FACTORIAL
MICROPY_OPT_MPZ_BITWISE
MICROPY_PY_BUILTINS_EXECFILE
MICROPY_PY_FRAMEBUF
MICROPY_PY_MATH_CONSTANTS
MICROPY_PY_MATH_FACTORIAL
MICROPY_PY_SYS_ATTR_DELEGATION
MICROPY_PY_SYS_PS1_PS2
MICROPY_PY_SYS_STDIO_BUFFER
MICROPY_PY_UASYNCIO
MICROPY_PY_URANDOM_EXTRA_FUNCS
MICROPY_PY_URE_SUB
MICROPY_REPL_EMACS_EXTRA_WORDS_MOVE
MICROPY_REPL_EMACS_WORDS_MOVE
MICROPY_VFS_FAT
MICROPY_VFS_LFS1
MICROPY_VFS_LFS2
```

jimmo committed:
ports/unix/variants: Remove freedos and fast variants.

freedos is untested by CI. fast is not a good name for what it does.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
ports/unix: Refactor mpconfigport.h and mpconfigvariant.h.

This is a no-op for coverage and minimal.

standard and dev have been merged and enable the same feature set as
a typical bare-metal board. Remove the CI to build the dev build.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
py/mpconfig: Make feature levels available to mpconfigport.h.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
py/mpconfig.h: Add "everything" features from coverage.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
ports/unix: Refactor mpconfigvariant.mk.

All variants (except minimal) enable text compression and fat/lfs, so move
them to the common mpconfigport.mk.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

andrewleech commented on ports/unix/variants/minimal/mpconfigvariant.h:
```diff
@@ -26,118 +26,44 @@
 
-// Prevent the rest of the default mpconfigport.h being used.
-#define MICROPY_UNIX_MINIMAL (1)
+#define MICROPY_CONFIG_ROM_LEVEL (MICROPY_CONFIG_ROM_LEVEL_MINIMUM)
```
I would suggest renaming this variant from `minimal` to `minimum` to match the rom level name.

jimmo replied:
It's the "minimum feature set" but the "minimal build".

Well, that was my logic at the time :) I'm happy to rename.

andrewleech replied:
I've thought / discussed in the past it'd be nice to have a variant that matches each of the rom levels, seems like this might be a good time to do it :-)

jimmo replied:
Can do this in a separate PR.

andrewleech commented on ports/unix/variants/standard/mpconfigvariant.h:
```diff
@@ -27,18 +27,5 @@
 // Set base feature level.
 #define MICROPY_CONFIG_ROM_LEVEL (MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES)
```
Similarly, could be less confusing in the long run if this variant was called "extra"?

jimmo replied:
Not sure... It's the standard (i.e. default) variant, it just happens to use the "extra" feature set.

andrewleech commented:
Everything you've written above @jimmo makes a lot of sense to me, thanks for this!

I agree that it makes sense to match the embedded ones re: select / working asyncio etc.

For clarity, we could have a variant for each of the defined ROM levels, that could be directly useful for people who want to use this port for testing code for compatibility with their particular target chip's level.

My only worry is that considering how "normal" the use of `dev` is currently in the wild this could be quite a breaking change for a lot of people. Would it be worth keeping the `dev` folder around but have it directly import from `standard` (or `extra` if the variant is named to match the rom level)?

jimmo commented:
> My only worry is that considering how "normal" the use of dev is currently in the wild this could be quite a breaking change for a lot of people. Would it be worth keeping the dev folder around but have it directly import from standard (or extra if the variant is named to match the rom level)?

I forgot to mention, part of the motivation for this PR was that we just did a workflow break with the change to the output binary path. So figured I'd get this in at the same time.

andrewleech commented:
> I forgot to mention, part of the motivation for this PR was that we just did a workflow break with the change to the output binary path. So figured I'd get this in at the same time.

Yeah fair point. I had seen that commit and flagged it as something we'd have to deal with in our scripts at work, it does make sense to go for gold here in one release!

jimmo committed:
py/mpconfig: Add LFS1/LFS2 options to match FAT/posix.

Also fixes the #ifndef for FAT & posix.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Rebased.

Minor note -- I missed `MICROPY_VFS_LFS1` and `MICROPY_VFS_LFS2` in the list of new things which are now enabled in the standard build (along with `MICROPY_VFS_FAT`). This is because there is no place in the codebase that does `#define MICROPY_VFS_LFS1`, so my script to find all of these options didn't see it. Added a commit to add these to mpconfig.h.

jimmo committed:
tests/extmod/uasyncio_heaplock.py: Force SKIP on stackless.

This is a latent issue that wasn't caught by CI because there was no
configuration that had both stackless+uasyncio.

The previous check to skip with stackless builds only worked when the
bytecode emitter was used by default. Force the check to use the bytecode
emitter.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Added a fix for the settrace_stackless build.

dpgeorge commented:
Merged in 89a0fefb6c5c730a7b740cf31e44a6c76c3993b1 through 6c376a93067230b3b9d7d4adc1b804df18b12dcd
